ko/lessons/specifics/debugging.md

```diff
@@ -0,0 +1,195 @@
+
+```elixir
+defp deps do
+  [{:dialyxir, "~> 0.4", only: [:dev]}]
+end
+```
+
+그리고 다음을 실행하세요.
+
+```shell
+$ mix deps.get
+...
+$ mix deps.compile
+```
+
+첫 번째 명령어는 Dialyxir를 다운로드하고 설치할 것입니다. Hex도 같이 설치할 것인지 물어볼 수도 있습니다. 두 번째 명령어는 Dialyxir 애플리케이션을 컴파일할 것 입니다. Dialyxir를 전역적으로 설치하고 싶으면, [문서](https://github.com/jeremyjh/dialyxir#installation)를 읽어보세요.
+
+마지막으로 Dialyzer를 실행해 PLT(Persistent Lookup Table)를 다시 빌드합니다. Erlang이나 Elixir의 새버전을 설치할 때마다 이작업이 필요합니다. 다행이도, Dialyzer는 사용할 때마다 표준 라이브러리를 분석 하려하지 않습니다. 전부 다운로드하는데에는 조금 시간이 걸립니다.
+
+```shell
+$ mix dialyzer --plt
+Starting PLT Core Build ... this will take awhile
+dialyzer --build_plt --output_plt /.dialyxir_core_18_1.3.2.plt --apps erts kernel stdlib crypto public_key -r /Elixir/lib/elixir/../eex/ebin /Elixir/lib/elixir/../elixir/ebin /Elixir/lib/elixir/../ex_unit/ebin /Elixir/lib/elixir/../iex/ebin /Elixir/lib/elixir/../logger/ebin /Elixir/lib/elixir/../mix/ebin
+  Creating PLT /.dialyxir_core_18_1.3.2.plt ...
+...
+ done in 5m14.67s
+done (warnings were emitted)
+```
+
+## 코드의 정적 분석
+
+이제 Dialyxir를 사용할 준비가 되었습니다.
+
+```shell
+$ mix dialyzer
+...
+examples.ex:3: Invalid type specification for function 'Elixir.Examples':sum_times/1. The success typing is (_) -> number()
+...
+```
+
+Dialyzer의 메세지는 명확합니다. `sum_times/1`함수의 반환값이 선언과 다르다고 합니다. 이것은 `Enum.sum/1`이 `integer`가 아닌 `number`를 반환하지만 `sum_times/1`은 `integer` 를 반환하기 때문입니다.
+
+`number`는 `integer`가 아니기 때문에 에러가 발생합니다. 어떻게 고칠 수 있을까요? `number`를 `integer`로 바꾸기위해 `round/1`을 사용할 필요가 있습니다.
+
+```elixir
+@spec sum_times(integer) :: integer
+def sum_times(a) do
+    [1, 2, 3]
+    |> Enum.map(fn el -> el * a end)
+    |> Enum.sum
+    |> round
+end
+```
+
+확인해 봅시다.
+
+```shell
+$ mix dialyzer
+...
+  Proceeding with analysis... done in 0m0.95s
+done (passed successfully)
+```
+
+사양과 정적 분석 도구로 코드를 스스로 테스트하고 버그가 적도록 만들 수 있습니다.
+
+# 디버깅
+
+코드의 정적분석만으로 충분하지 않을 때가 있습니다. 버그를 찾기 위해 실행 흐름을 이해할 필요가 있을 수도 있습니다. 이것을 하는 가장 쉬운 방법은 코드의 흐름과 값을 확인하기 위해 `IO.puts/2`같은 출력문을 사용하는 것이지만, 이 기술은 원시적이고 한계가 있습니다. 고맙게도 Elixir 코드를 디버깅하기위해 Erlang의 디버거를 사용할 수 있습니다.
+
+기초적인 모듈이 있다고 합시다.
+
+```elixir
+defmodule Example do
+
+  def cpu_burns(a, b, c) do
+    x = a * 2
+    y = b * 3
+    z = c * 5
+
+    x + y + z
+  end
+
+end
+```
+
+
+`iex`를 실행합니다.
+
+```bash
+$ iex -S mix
+```
+
+그리고 디버거를 실행합니다.
+
+```elixir
+iex > :debugger.start()
+{:ok, #PID<0.307.0>}
+```
+
+Erlang `:debugger` 모듈로 디버거에 접근할 수 있습니다. `start/1` 함수를 사용해 설정할 수 있습니다.
+
++ 파일 경로를 넘기면 외부 설정 파일을 사용할 수 있습니다.
++ 인자가 `:local`나 `:global`이라면 디버거는 다음과 같이 동작합니다.
+    + `:global` – 디버거는 모든 알려진 노드에 대해 코드를 중단합니다. 기본값입니다.
+    + `:local` – 디버거는 현제 노드에 대해서만 코드를 중단합니다.
+
+다음 단계에서 디버거에 위의 모듈을 붙여보겠습니다.
+
+```elixir
+iex > :int.ni(Example)
+{:module, Example}
+```
+
+`:int` 모듈은 브레이크 포인트를 만들고 코드 실행 단계를 따라갈 수 있게하는 인터프리터입니다.
+
+디버거를 시작하면 이런 창을 볼 수 있습니다.
+
+![Debugger Screenshot 1]({{ site.url }}/assets/debugger_1.png)
+
+모듈을 디버거에 붙인 다음 왼쪽에 있는 메뉴에서 사용 가능합니다.
+
+![Debugger Screenshot 2]({{ site.url }}/assets/debugger_2.png)
+
+## 브레이크 포인트 만들기
```

Comment:
Since there are many (and common) preferences how to transalte that, let's keep this and wait for somebody who has a strong preference among them.